spanner/request_id_header_test.go

```diff
@@ -0,0 +1,1278 @@
+				{ProcessID: procID, ClientID: 1, ChannelID: 4, RequestNo: 3, RPCNo: 4},
+			},
+			wantErr: "",
+		},
+		{
+			name: "Same clientID, prev has higher RPCNo",
+			in: []*requestIDSegments{
+				{ProcessID: procID, ClientID: 1, ChannelID: 1, RequestNo: 1, RPCNo: 2},
+				{ProcessID: procID, ClientID: 1, ChannelID: 1, RequestNo: 1, RPCNo: 1},
+			},
+			wantErr: "sameRequestNo yet #[0].RPCNo=2 >= #[1].RPCNo=1",
+		},
+		{
+			name: "Same clientID, same channelID, prev has lower RPCNo",
+			in: []*requestIDSegments{
+				{ProcessID: procID, ClientID: 1, RequestNo: 2, ChannelID: 1, RPCNo: 1},
+				{ProcessID: procID, ClientID: 1, RequestNo: 2, ChannelID: 1, RPCNo: 2},
+			},
+			wantErr: "",
+		},
+		{
+			name: "Same clientID, prev has higher clientID",
+			in: []*requestIDSegments{
+				{ProcessID: procID, ClientID: 2, RequestNo: 1, RPCNo: 1, ChannelID: 1},
+				{ProcessID: procID, ClientID: 1, RequestNo: 1, RPCNo: 1, ChannelID: 1},
+			},
+			wantErr: "", // Requests can execute in parallel.
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// 1. Each segment but be valid!
+			err := ensureMonotonicityOfRequestIDs(tt.in)
+			if tt.wantErr != "" {
+				if err == nil {
+					t.Fatal("Expected a non-nil error")
+				}
+				if !strings.Contains(err.Error(), tt.wantErr) {
+					t.Fatalf("Error mismatch\n\t%q\ncould not be found in\n\t%q", tt.wantErr, err)
+				}
+				return
+			}
+
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+		})
+	}
+}
+
+func (it *interceptorTracker) unaryClientInterceptor(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	it.nUnaryClientCalls.Add(1)
+	reqID, err := checkForMissingSpannerRequestIDHeader(opts)
+	if err != nil {
+		return err
+	}
+
+	it.mu.Lock()
+	it.unaryClientRequestIDSegments = append(it.unaryClientRequestIDSegments, reqID)
+	it.mu.Unlock()
+
+	// fmt.Printf("unary.method=%q\n", method)
+	// fmt.Printf("method=%q\nReq: %#v\nRes: %#v\n", method, req, reply)
+	// Otherwise proceed with the call.
+	return invoker(ctx, method, req, reply, cc, opts...)
+}
+
+func (it *interceptorTracker) streamClientInterceptor(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	it.nStreamClientCalls.Add(1)
+	reqID, err := checkForMissingSpannerRequestIDHeader(opts)
+	if err != nil {
+		return nil, err
+	}
+
+	it.mu.Lock()
+	it.streamClientRequestIDSegments = append(it.streamClientRequestIDSegments, reqID)
+	it.mu.Unlock()
+
+	// fmt.Printf("stream.method=%q\n", method)
+	// Otherwise proceed with the call.
+	return streamer(ctx, desc, cc, method, opts...)
+}
+
+func newInterceptorTracker() *interceptorTracker {
+	return &interceptorTracker{
+		nUnaryClientCalls:  new(atomic.Uint64),
+		nStreamClientCalls: new(atomic.Uint64),
+	}
+}
+
```

Comment:
The tests that I would have expected and that are missing here are:

### Base Scenario
Verify that we see exactly the request IDs that we expect when we have a scenario that can be 100% predicted. That would be for example:
1. A Client that has a SessionPoolConfig with Min/Max: 1 (this prevents concurrent BatchCreateSessions calls from making things unpredictable)
2. The application waits until the session pool has been filled with 1 session.
3. The client then executes a read-only transaction that executes 1 query.

Then verify exactly the request-IDs that we see (so not only that they are monotonically increasing)

### Base Unary Retry
Same as above, but let one of the unary RPCs be retried by simulating an UNAVAILABLE error.

### Base Streaming Retry
Same as above, but let the streaming ExecuteStreamingSql be retried by simulating an UNAVAILABLE error.

### Session Pool Growth
I think that there's a bug in the current assignment of channel IDs, as the session pool creates a new vkit.Client when it grows the pool. That does not create new gRPC channels, but my impression is that this event leads to a new channel-ID being assigned to the new session(s).

Reply:
When I made updates to the change to infer the channelID directly from grpc.ClientConn from ConnPool, that eliminated that possibility but indeed, let me add a test to grow the session pool and test out channelID doling out.